Modules/_ctypes/_ctypes.c

```diff
@@ -5150,30 +5150,36 @@ Pointer_get_contents(CDataObject *self, void *closure)
 
     stgdict = PyObject_stgdict((PyObject *)self);
     assert(stgdict); /* Cannot be NULL for pointer instances */
+    assert(stgdict->proto);
 
-    keep = GetKeepedObjects(self);
-    if (keep != NULL) {
-        // check if it's a pointer to a pointer:
-        // pointers will have '0' key in the _objects
-        ptr_probe = PyDict_GetItemString(keep, "0");
-
-        if (ptr_probe != NULL) {
-            ptr2ptr = (CDataObject*) PyDict_GetItemString(keep, "1");
-            if (ptr2ptr ==  NULL) {
-                PyErr_SetString(PyExc_ValueError,
-                "Unexpected NULL pointer in _objects");
-                return NULL;
-            }
-            // don't construct a new object,
+    if (self->b_objects != NULL && PyDict_CheckExact(self->b_objects)) {
+        // Pointer_set_contents uses KeepRef(self, 1, value); we retrieve that
+        ptr2ptr = PyDict_GetItemString(self->b_objects, "1");
+        if (ptr2ptr ==  NULL) {
+            PyErr_SetString(PyExc_ValueError,
+            "Unexpected NULL pointer in _objects");
+            return NULL;
+        }
+        // if our base pointer is cast from another type,
+        // its `_type_` proto will be incompatible with the
+        // type of the object stored in `b_objects["1"]` because
+        // `_objects` is shared between casts and the original.
+        int res = PyObject_IsInstance(ptr2ptr, stgdict->proto);
+        if (res == -1) {
+            return NULL;
```

Comment:
Do you think it might be useful to call `PyErr_SetString` here?

Reply:
I did what other code in `_ctypes.c` did for this usage, which wasn't to set the error string.

Reply:
Oh, yes, `PyObject_IsInstance` will do it if there's an error